content/en/blog/2023/2023-03-08-scheduler-plugin-reusing/index.md

```diff
@@ -0,0 +1,161 @@
+
+The dependency to the composite does require some understanding of the composite code, and there may be some introduced fragility with having that dependency. For example, looking into the _NodeResourceFit_ plugin internals reveals that it utilizes the [framework.CycleState](https://github.com/kubernetes/kubernetes/blob/2f977fd8c400971deed365dcc437e519bce475f0/pkg/scheduler/framework/cycle_state.go#L42-L47) during running of other Plugin interfaces. Luckily during running of the `ScorePlugin` interface parts no such state is utilized, so in this case it wasn't necessary to add and run those other plugin interfaces. However should such a thing change in the composite plugin, then running of the composite using plugin might also require changes.
+
+### More efficient and flexible alternative, plugin pipelining
+
+Going through the exercise of building a new plugin via composition and realizing the caveats of such reuse raised some questions related to if and how things could be better in an ideal world. Such as:
+
+* Could a plugin utilize scores from multiple other plugins?
+* Could multiple new plugins utilize scores from the same plugin, without running the same scoring logic N times?
+* Could the dependency to the reused plugin be made less fragile, reducing the coupling?
+* Could reuse be achieved without duplicated args and configuring same algorithms the same way, twice (or more)?
+* Could things be configured more dynamically instead of hard-coding to the sources one plugin uses another plugin?
+
+These questions had the author thinking about finding a better solution than just composition. Also the documentation of the [framework.CycleState](https://github.com/kubernetes/kubernetes/blob/2f977fd8c400971deed365dcc437e519bce475f0/pkg/scheduler/framework/cycle_state.go#L42-L47) was looking rather tempting.
+
+> **Warning**
+> The following section is speculative and based on proof-of-concepting by the author. It doesn't represent what you can do with the Kubernetes scheduler at the time of writing. Some day the concept below might be part of Kubernetes, should the PoC develop into a successful KEP. For now, this is merely a sig-scheduling discussion item which shows the potential of the CycleState.
+
+Ideally, a new plugin could ask for getting the scores from another plugin, or multiple plugins, without rerunning them. The tricky thing is, that plugins run highly parallel in order to execute faster. Thus proper synchronization would be needed. Also it would be ideal, if no changes to existing plugins would be required, and the framework would provide for plugins in need of reusing other plugins' results.
+
+This got the author thinking, that if the `framework.CycleState` object could contain a channel for the scores, and if the framework would create such a channel when it is needed, then the parallel running plugins which are interested in score reuse, could wait for the scores to appear from such "ScoreChannels".
```

Comment:
```suggestion
This got me thinking that if the `framework.CycleState` object could contain a channel for the scores, and if the framework would create such a channel when it is needed, then the parallel running plugins which are interested in score reuse could wait for the scores to appear from such `ScoreChannels`.
```

even better rewritten as shorter sentences (commas are not enough; a rewrite would use more periods)